File: week32/키패드 누르기_Nunu-0.py
[프로그래머스] 키패드 누르기
def solution(numbers, hand):
    answer = ''
    leftN = [1, 4, 7] # 왼쪽열의 숫자
    rightN = [3, 6, 9] # 오른쪽열의 숫자
    leftH = 10 # 왼손 위치(*)
    rightH = 12 # 오른손 위치(#)
    
    for num in numbers:
        # 왼쪽열의 숫자
        if num in leftN:
            answer += 'L'
            leftH = num
        # 오른쪽열의 숫자
        elif num in rightN:
            answer +='R'
            rightH = num
        else:
            # 숫자가 0일 때 위치 11로 재설정
            if num == 0:
                num = 11
            
            # 숫자와 양 손 간의 간격차 
            leftGap = abs(num - leftH)
            rightGap = abs(num - rightH)
            # 거리
            leftD = leftGap % 3 + leftGap // 3
            rightD = rightGap % 3 + rightGap // 3
            '''
            % = 좌우간의 간격
            // = 상하의 간격
            '''
            # 왼손이 숫자와 더 가깝다면
            if leftD < rightD:
                answer += 'L'
                leftH = num
            # 오른손이 숫자와 더 가깝다면
            elif leftD > rightD:
                answer += 'R'
                rightH = num
            # 양손의 거리가 같다면 
            elif leftD == rightD:
                if hand == 'left':
                    answer += 'L'
                    leftH = num
                else:
                    answer += 'R'
                    rightH = num         
            
    return answer

File: week40/키패드 누르기_Nunu-0.js
// 키패드 누르기 (22.07.01)
// https://programmers.co.kr/learn/courses/30/lessons/67256?language=javascript
function solution(numbers, hand) {
    let answer = '';
    const Lnum = [1, 4, 7];
    const Rnum = [3, 6, 9];
    const Cnum = [2,5,8,11];
    let Lpos = 10; // 왼손 위치(*)
    let Rpos = 12; // 오른손 위치(#)
    
    numbers.forEach(n =>{
        if (Lnum.includes(n)){
            answer += 'L';
            Lpos = n;
        }
        else if(Rnum.includes(n)){
            answer += 'R';
            Rpos = n;
        }
        else{
            if (n == 0) n = 11;
            
            const Ldis = Cnum.includes(Lpos)?Math.abs(n - Lpos)/3:Math.abs(n-Lpos-1)/3+1;
            const Rdis = Cnum.includes(Rpos)?Math.abs(n - Rpos)/3:Math.abs(n-Rpos+1)/3+1;   
            console.log(Ldis, Rdis)
            if (Ldis < Rdis){
                answer += 'L';
                Lpos = n;
            }
            else if (Ldis > Rdis) {
                answer += 'R';
                Rpos = n;
            }
            else{
                if (hand == 'left'){
                    answer += 'L';
                    Lpos = n;
                }
                else{
                    answer += 'R';
                    Rpos = n;
                }
            }
        }
    });
    return answer
}